hail/src/main/scala/is/hail/expr/ir/IsPure.scala

```diff
@@ -0,0 +1,11 @@
+package is.hail.expr.ir
+
+import is.hail.types.virtual.TVoid
+
+object IsPure {
+  def apply(x: IR): Boolean = x match {
```

Comment:
`Die`?

Reply:
We don't actually treat Die as side effecting. For instance, in
```
GetField(
  MakeStruct(foo = I32(5), bar = Die())
  "foo")
```
we will prune "bar", even though it changes the program from dying to returning a value.

Formally, our semantics is slightly non-deterministic: there is at most one correct value result, and any number of correct exception results, and running the program is allowed to produce any of those behaviors.

Reply:
That seems like a bug to me

Reply:
Only if it violates the semantics of the given IR. We intentionally use a semantics for which this is not a bug. Preserving all potential errors would be a huge burden on the optimizer. For instance, one important optimization is pushing computation to happen after a filter. But if the computation could possibly error (eg divide by zero), then that's an invalid optimization. Or if a big struct is computed, then only some of its fields are used, we'd be forced to compute all the unused fields in case they error.